primo/utils/clustering_utils.py

```diff
@@ -97,39 +101,67 @@ def distance_matrix(wd: WellData, weights: dict) -> np.ndarray:
     )
 
 
-def perform_clustering(wd: WellData, distance_threshold: float = 10.0):
+def _well_clusters(wd: WellData) -> dict:
```

Comment:
As an overall comment, it is important to think of what can be "generalized" vs what needs to be specialized for the PRIMO project. I could foresee a scenario where we use Louvain clustering for something else in the future. In that case, having it generally implemented (instead of directly tied into 'WellData' could be useful.

No changes required at this time, something to think about for the future